src/disco/topo/fd_topob.c

```diff
@@ -392,6 +393,95 @@ validate( fd_topo_t const * topo ) {
   }
 }
 
+void
+fd_topob_auto_layout( fd_topo_t * topo ) {
+  /* Incredibly simple automatic layout system for now ... just assign
+     tiles to CPU cores in NUMA sequential order, except for a few tiles
+     which should be floating. */
+
+  char const * FLOATING[] = {
+    "metric",
+    "bencho",
+    "bhole",  // FIREDANCER only
+  };
+
+  char const * ORDERED[] = {
+    "benchg",
+    "benchs",
+    "net",
+    "quic",
+    "verify",
+    "dedup",
+    "pack",
+    "bank",
+    "poh",
+    "pohi",   // FIREDANCER only
+    "shred",
+    "store",
+    "storei", // FIREDANCER only
+    "sign",
+    "gossip", // FIREDANCER only
+    "repair", // FIREDANCER only
+    "replay", // FIREDANCER only
+    "thread", // FIREDANCER only
+    "sender", // FIREDANCER only
+  };
+
+  for( ulong i=0UL; i<topo->tile_cnt; i++ ) {
+    fd_topo_tile_t * tile = &topo->tiles[ i ];
+    tile->cpu_idx = ULONG_MAX;
+  }
+
+  ulong cpu_ordering[ USHORT_MAX ] = { 0UL };
+  ulong num_cpus = fd_numa_cpu_cnt();
+
+  ulong next_cpu_idx = 0UL;
+  for( ulong i=0UL; i<fd_numa_node_cnt(); i++ ) {
+    for( ulong j=0UL; j<num_cpus; j++ ) {
+      ulong numa_node = fd_numa_node_idx( j );
+      if( FD_UNLIKELY( numa_node!=i ) ) continue;
+      cpu_ordering[ next_cpu_idx++ ] = j;
```

Comment:
Make it abort if it goes past the end of the array instead of being an OOB write

Reply:
You can do this in O(n) with two passes (count CPUs per numa node, prefix sums, then assign) but it doesn't matter since this is startup and the number of CPUs/numa nodes is small